software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java

```diff
@@ -379,20 +418,30 @@ public MachineLocation call() throws Exception {
             if (!(location instanceof LocalhostMachineProvisioningLocation))
                 log.info("Starting {}, obtaining a new location instance in {} with ports {}", new Object[]{entity(), location, flags.get("inboundPorts")});
             entity().sensors().set(SoftwareProcess.PROVISIONING_LOCATION, location);
+            Transition expectedState = entity().sensors().get(Attributes.SERVICE_STATE_EXPECTED);
+            
+            // BROOKLYN-263: see corresponding code in doStop()
+            if (expectedState != null && (expectedState.getState() == Lifecycle.STOPPING || expectedState.getState() == Lifecycle.STOPPED)) {
+                throw new IllegalStateException("Provisioning aborted before even begun for "+entity()+" in "+location+" (presumably by a concurrent call to stop");
+            }
+            entity().sensors().set(PROVISIONING_TASK_STATE, ProvisioningTaskState.RUNNING);
+            
             MachineLocation machine;
             try {
                 machine = Tasks.withBlockingDetails("Provisioning machine in " + location, new ObtainLocationTask(location, flags));
-                if (machine == null)
-                    throw new NoMachinesAvailableException("Failed to obtain machine in " + location.toString());
-            } catch (Exception e) {
-                throw Exceptions.propagate(e);
+                entity().sensors().set(PROVISIONED_MACHINE, machine);
```

Comment:
I'm willing to take that risk - the description says "internal" etc. The problem is that otherwise there is a big gap between when provisioning returns and when we add the machine to the location. I worry that start() will not always go through all those steps if there has been a concurrent call to stop. It might skip executing those subsequent tasks. With the change I made, we set DONE and the machine in a finally block so it removes that risk.